src/java.base/aix/classes/sun/nio/ch/PollsetPoller.java

```diff
@@ -25,39 +26,86 @@
 package sun.nio.ch;
 
 import java.io.IOException;
+import java.time.Instant;
+import sun.nio.ch.Pollset;
 
 /**
  * Poller implementation based on the AIX Pollset library.
  */
 
 class PollsetPoller extends Poller {
 
+    static { Pollset.init(); /* Dynamically loads pollset C functions */ }
+
+    private final int setid;
+    private int setsize;
+
     PollsetPoller(boolean read) throws IOException {
         super(read);
+        this.setsize = 0;
+        this.setid = Pollset.pollsetCreate();
     }
 
     @Override
     int fdVal() {
-        // Stub
-        throw new UnsupportedOperationException("Unimplemented on AIX");
+        return setid;
     }
 
     @Override
-    void implRegister(int fdVal) throws IOException {
-        // Stub
-        throw new UnsupportedOperationException("Unimplemented on AIX");
+    void implRegister(int fd) throws IOException {
+        int ret = Pollset.pollsetCtl(setid, Pollset.PS_MOD, fd,
+        Pollset.PS_POLLPRI | (this.reading() ? Net.POLLIN : Net.POLLOUT));
+        if (ret != 0) {
+            throw new IOException("Unable to register fd " + fd);
+        }
+        setsize++;
     }
 
     @Override
-    void implDeregister(int fdVal) {
-        // Stub
-        throw new UnsupportedOperationException("Unimplemented on AIX");
+    void implDeregister(int fd) {
+        assert (setsize > 0);
+        int ret = Pollset.pollsetCtl(setid, Pollset.PS_DELETE, fd, 0);
+        assert ret == 0;
+        setsize--;
     }
 
+    /**
+      * Main poll method. The AIX Pollset library does not appear to pick up changes to the pollset
+      * (the set of fds being polled) while blocked on a call to this method. These changes happen
+      * regularly in the poll-loop thread and update thread from Poller.java.
+      * To address this difficulty, we break poll calls into 100ms sub-calls and emulate the timout.
+      */
     @Override
     int poll(int timeout) throws IOException {
-        // Stub
-        throw new UnsupportedOperationException("Unimplemented on AIX");
+        int n;
+        switch (timeout) {
+            case 0:
+                n = pollInner(0);
+                break;
+            case Pollset.PS_NO_TIMEOUT:
+                do { n = pollInner(100); } while (n == 0);
+                break;
+            default:
+                Instant end = Instant.now().plusMillis(timeout);
+                do { n = pollInner(100); } while (n == 0 && Instant.now().isBefore(end));
```

Comment:
Now, the Poller uses poll(-1) to poll indefinitely so the 0/default cases aren't used. If we do start to use the timed case then L90 probably should probably be optimized to avoid Instant.now. Also, probably should be reformatted to make the do-while loop a bit easier to read.

Reply:
I'm happy to tweak this case. In terms of formatting, would you prefer?

```
do { n = pollInner(100); }
while (n == 0 && Instant.now().isBefore(end));
```

or the more conventional

```
do {
  n = pollInner(100);
} while (n == 0 && Instant.now().isBefore(end));
```

or another idea?

---

In regards to `Instant.now(...)`. I'm not sure I understand your suggestion. Could you please clarify? Is there a different time-object I should use instead?

Reply:
Just waiting for clarification from you on this discussion @AlanBateman.